actors/miner/tests/util.rs

```diff
@@ -1682,12 +1684,452 @@ pub struct DeadlineStateSummary {
+                    0
+                }
+            };
+
+        let partition = partition.clone();
+        PartitionStateSummary {
+            all_sectors: partition.sectors,
+            live_sectors: live,
+            faulty_sectors: partition.faults,
+            recovering_sectors: partition.recoveries,
+            unproven_sectors: partition.unproven,
+            terminated_sectors: partition.terminated,
+            live_power,
+            active_power,
+            faulty_power: partition.faulty_power,
+            recovering_power: partition.recovering_power,
+            expiration_epochs,
+            early_termination_count,
+        }
+    }
+}
+
+#[derive(Default)]
+struct ExpirationQueueStateSummary {
+    on_time_sectors: BitField,
+    early_sectors: BitField,
+    #[allow(dead_code)]
+    active_power: PowerPair,
+    #[allow(dead_code)]
+    faulty_power: PowerPair,
+    #[allow(dead_code)]
+    on_time_pledge: TokenAmount,
+    expiration_epochs: Vec<ChainEpoch>,
+}
+
+impl ExpirationQueueStateSummary {
+    // Checks the expiration queue for consistency.
+    fn create<BS: Blockstore>(
+        expiration_queue: &ExpirationQueue<BS>,
+        live_sectors: &SectorsMap,
+        partition_faults: &BitField,
+        quant: QuantSpec,
+        sector_size: SectorSize,
+        acc: &mut MessageAccumulator,
+    ) -> Self {
+        let mut seen_sectors: HashSet<SectorNumber> = HashSet::new();
+        let mut all_on_time: Vec<BitField> = Vec::new();
+        let mut all_early: Vec<BitField> = Vec::new();
+        let mut expiration_epochs: Vec<ChainEpoch> = Vec::new();
+        let mut all_active_power = PowerPair::zero();
+        let mut all_faulty_power = PowerPair::zero();
+        let mut all_on_time_pledge = BigInt::zero();
+        let mut first_queue_epoch: ChainEpoch = -1;
+
+        let ret = expiration_queue.amt.for_each(|epoch, expiration_set| {
+            let epoch = epoch as i64;
+            let mut acc = acc.with_prefix(&format!("expiration epoch {epoch}: "));
+            let quant_up = quant.quantize_up(epoch);
+            acc.require(quant_up == epoch, &format!("expiration queue key {epoch} is not quantized, expected {quant_up}"));
+            if first_queue_epoch == -1 {
+                first_queue_epoch = epoch;
+            }
+
+            expiration_epochs.push(epoch);
+
+            let mut on_time_sectors_pledge = BigInt::zero();
+            for sector_number in expiration_set.on_time_sectors.iter() {
+                // check sectors are present only once
+                if !seen_sectors.insert(sector_number) {
+                    acc.add(&format!("sector {sector_number} in expiration queue twice"));
+                }
+
+                // check expiring sectors are still alive
+                if let Some(sector) = live_sectors.get(&sector_number) {
+                    // The sector can be "on time" either at its target expiration epoch, or in the first queue entry
```

Comment:
This is slightly out of date.  Since snap deals we have deprecated the old CC replaced mechanism, so it is now an error to see an on time sector anywhere but its target expiration epoch.  Drop `epoch == first_queue_epoch` and all accounting of `first_queue_epoch` can now go away.

Reply:
Nice, didn't know about this. Thanks.